zerver/actions/reminders.py

```diff
@@ -0,0 +1,122 @@
+
+from zerver.actions.message_send import check_message, internal_send_private_message
+from zerver.actions.scheduled_messages import do_schedule_messages
+from zerver.lib.addressee import Addressee
+from zerver.lib.exceptions import JsonableError
+from zerver.lib.message import access_message
+from zerver.lib.url_encoding import huddle_narrow_url, personal_narrow_url, topic_narrow_url
+from zerver.models import (
+    Client,
+    Message,
+    ScheduledMessage,
+    Stream,
+    UserProfile,
+    get_display_recipient,
+    get_system_bot,
+)
+
+
+# This function is similar to one used in fenced_code.ts
+def get_unused_fence(content: str) -> str:
+    # Define the regular expression pattern to match ``` fences
+    fence_length_re = re.compile(r"^ {0,3}(`{3,})", re.MULTILINE)
+
+    # Initialize the length variable to 3, corresponding to default fence length
+    length = 3
+    matches = fence_length_re.findall(content)
+    for match in matches:
+        length = max(length, len(match) + 1)
+
+    return "`" * length
+
+
+def by_conversation_and_time_url(message: Message, current_user: UserProfile) -> str:
+    if message.is_stream_message():
+        stream_id = message.recipient.type_id
+        narrow_link = topic_narrow_url(
+            realm=message.sender.realm,
+            stream=Stream.objects.get(id=stream_id),
+            topic=message.topic_name(),
+        )
+    elif message.recipient.type == message.recipient.PERSONAL:
+        narrow_link = personal_narrow_url(
+            realm=message.sender.realm,
+            sender=message.sender,
+        )
+    else:
+        narrow_link = huddle_narrow_url(
+            user=current_user,
+            display_recipient=get_display_recipient(message.recipient),
+        )
+
+    return f"{narrow_link}/near/{message.id}"
+
+
+def get_formatted_content_from_message(message: Message, current_user: UserProfile) -> str:
+    if message.is_stream_message():
+        stream_id = message.recipient.type_id
+        stream = Stream.objects.get(id=stream_id)
+        narrow_link = topic_narrow_url(
+            realm=message.sender.realm,
+            stream=Stream.objects.get(id=stream_id),
+            topic=message.topic_name(),
+        )
+        content = "You requested a reminder for the following message sent to "
+        content += f"[{stream.name} > {message.topic_name()}]({narrow_link})."
+    else:
+        content = "You requested a reminder for the following private message."
+    content += "\n\n"
+    content += _("@_**{sender_full_name}|{sender_id}** [said]({conversation_url}):").format(
+        sender_full_name=message.sender.full_name,
+        sender_id=message.sender.id,
+        conversation_url=by_conversation_and_time_url(message, current_user),
+    )
+    content += "\n"
+    fence = get_unused_fence(content)
+    quoted_message = f"{fence}quote\n{message.content}\n{fence}"
+    content += quoted_message
+    return content
```

Comment:
We probably need to do some changes to make this handle reminders of messages that were exactly 10K characters long, to avoid having our closing fence be truncated? Idunno, might be OK given that we use an unused fence and our Markdown implementation will assume a final closing fence, but worth testing.

Reply:
It seems it is still truncated to 10k characters and [message truncated] gets added at the end. Added a test case for this too. This seems fine to me as the link to the original message is still present here.

Reply:
And critically, the Markdown system still formats it correctly as quoted, right?